lib/compiler/container.ts

```diff
@@ -0,0 +1,161 @@
+      return this._units;
+    }
+    const containerDecorator = this.clazz.getDecorator(container.name);
+    if (containerDecorator && containerDecorator.isDecoratorFactory()) {
+      const options = containerDecorator.getArguments()[0];
+      if (!TypeGuards.isObjectLiteralExpression(options)) {
+        throw new Error('Invalid container decorator: ' + this.clazz.print());
+      }
+      const unitsProperty = options.getProperty('units');
+      if (!unitsProperty || !TypeGuards.isPropertyAssignment(unitsProperty)) {
+        throw new Error('Invalid container units: ' + this.clazz.print());
+      }
+      const unitsArray = unitsProperty.getInitializerIfKindOrThrow(
+        SyntaxKind.ArrayLiteralExpression
+      );
+      this._units = unitsArray.getElements().map(element => {
+        if (!TypeGuards.isIdentifier(element)) {
+          throw new Error(
+            'Invalid container unit must be identifier: ' + element.print()
+          );
+        }
+        const node = findDeclarationForIdentifier(element);
+        if (TypeGuards.isClassDeclaration(node)) {
+          return new UnitImpl(this, node);
+        } else {
+          throw new Error('Illegal node type for unit: ' + node.print());
+        }
+      });
+    } else {
+      this._units = [];
+    }
+    return this._units;
+  }
+
+  private get internals(): Component[] {
+    const pending = [
+      ...this.entryPoints.map(entryPoint => entryPoint.type),
+      ...this.units.map(unit => unit.unitComponent)
+    ];
+    const list: Component[] = [];
+
+    while (pending.length > 0) {
+      const component = pending.shift();
+      if (component) {
+        list.push(component);
+        pending.push(
+          ...component.constructorDependencies.map(
+            dependency => dependency.type
+          ),
+          ...component.propertyDependencies.map(dependency => dependency.type)
+        );
+      }
+    }
+    return list;
+  }
+
+  public get externals(): Component[] {
+    const filterInternals = (list: Component[]) =>
+      list.filter(component => !this.internals.includes(component));
+
+    const all = [
+      ...this.compiler.getComponents(this),
+      ...this.compiler.getLegacyComponents(this),
+      ...this.compiler.getLegacyFactories(this).map(info => info.component),
+      ...this.compiler.getLegacyExternals(this)
+    ];
+    if (this.compiler.getContainers().length > 1) {
+      const unassigned = filterInternals(
+        all.filter(component => !component.by)
+      );
+      if (unassigned.length > 0) {
+        throw new Error(
+          `Unassigned externals [${
+            unassigned[0].name
+          }] is not supported if using multiple containers`
+        );
+      }
+
+      return filterInternals(
+        all
+          .filter(component => component.by)
+          .filter(component => component.by!.clazz === this.clazz)
+      );
+    }
+    return filterInternals(all);
+  }
+
+  public get components(): Component[] {
```

Comment:
See my comments below.

I think the concept will get a lot easier now for most cases. This will reduce the entry barrier.
But in complex setups it is way more flexible compared to before.
TSDI still does not assume and structure or architecture besides that you need to use classes. :smile: That may be unpopular in the JS world, but I like it.